python_etl/churn.py

```diff
@@ -0,0 +1,654 @@
+        StructField('unique_domains_count',    DoubleType(),   True)
+    ])
+
+    def reduce_func(x, y):
+        return tuple(map(sum, zip(x, y)))
+
+    aggregated = countable.reduceByKey(reduce_func)
+
+    records_df = aggregated.map(lambda x: x[0] + x[1]).toDF(churn_schema)
+
+    # Apply some post-processing for other aggregates
+    # (i.e. unique_domains_count). This needs to be done when you want
+    # something other than just a simple sum
+    def average(total, n):
+        if not n:
+            return 0.0
+        return float(total) / n
+    average_udf = F.udf(average, DoubleType())
+
+    # Create new derived columns and any unecessary ones
+    records_df = (
+        records_df
+        # The total number of unique domains divided by the number of profiles
+        # over a set of dimensions. This should be aggregated using a weighted
+        # mean, i.e. sum(unique_domains_count_per_profile * n_profiles)
+        .withColumn('unique_domains_count_per_profile',
+                    average_udf(F.col('unique_domains_count'),
+                                F.col('n_profiles')))
+        # This value is meaningless because of overlapping domains between
+        # profiles
+        .drop('unique_domains_count')
+    )
+
+    return records_df
+
+
+def daterange(start_date, end_date):
+    for n in range(int((end_date - start_date).days) // 7):
+        yield (start_date + timedelta(n * 7)).strftime("%Y%m%d")
+
+
+def process_backfill(start_date, end_date, callback):
+    """ Import data from a start date to an end date.
+
+    :start_date ds: starting date in yyyymmdd
+    :end_date ds: ending date in yyymmdd
+    :callback: a callback accepting a datestring in format yyyymmdd
+    """
+    start_date = datetime.strptime(start_date, "%Y%m%d")
+    end_date = datetime.strptime(end_date, "%Y%m%d")
+    for d in daterange(start_date, end_date):
+        try:
+            callback(d)
+        except Exception as e:
+            print e
+
+
+def process_week(df, week_start, bucket, prefix):
+    result_df = compute_churn_week(df, week_start)
+
+    # Write to s3 as parquet, file size is on the order of 40MB. We
+    # bump the version number because v1 is the path to the old
+    # telemetry-batch-view churn data.
+    s3_path = "s3://{}/{}/week_start={}".format(bucket, prefix, week_start)
+    print("{}: Writing output as parquet to {}"
+          .format(datetime.utcnow(), s3_path))
+
+    result_df.repartition(1).write.parquet(s3_path, mode="overwrite")
+
+    print("Finished week {} at {}".format(week_start, datetime.utcnow()))
+
+
+@click.command()
+@click.option('-d', '--date', 'start_date',
+              default=fmt(datetime.now() - timedelta(7)),
+              help="datestring in form YYYYmmdd")
+@click.option('-b', '--bucket',
+              default='net-mozaws-prod-us-west-2-pipeline-analysis',
+              help="output s3 bucket")
+@click.option('-p', '--prefix', default='telemetry-test-bucket/churn-test',
```

Comment:
I don't think the defaults for `--bucket` and `--prefix` make sense. If you want to use `s3://telemetry-test-bucket`, then that should go into the `--bucket` argument.

In fact, do we even want defaults for these? Maybe it would be better not to have defaults and simply bail out if these args are not present (date, bucket, prefix).